espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/ComposeNodeFinder.kt

```diff
@@ -0,0 +1,89 @@
+import androidx.test.espresso.PerformException
+import android.view.View
+import android.widget.AdapterView
+import androidx.compose.ui.test.*
+
+import org.hamcrest.Description
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers
+import org.hamcrest.TypeSafeMatcher
+
+import java.util.ArrayList
+
+import io.appium.espressoserver.lib.handlers.exceptions.AppiumException
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidSelectorException
+import io.appium.espressoserver.lib.handlers.exceptions.XPathLookupException
+import io.appium.espressoserver.lib.model.Strategy
+import io.appium.espressoserver.lib.viewaction.ViewGetter
+
+import androidx.test.core.app.ApplicationProvider.getApplicationContext
+import androidx.test.espresso.AppNotIdleException
+import androidx.test.espresso.Espresso.onData
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.assertion.ViewAssertions.matches
+import androidx.test.espresso.matcher.ViewMatchers.isDescendantOfA
+import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
+import androidx.test.espresso.matcher.ViewMatchers.withClassName
+import androidx.test.espresso.matcher.ViewMatchers.withContentDescription
+import androidx.test.espresso.matcher.ViewMatchers.withId
+import androidx.test.espresso.matcher.ViewMatchers.withTagValue
+import androidx.test.espresso.matcher.ViewMatchers.withText
+import io.appium.espressoserver.EspressoServerRunnerTest
+import io.appium.espressoserver.lib.handlers.exceptions.InvalidElementStateException
+import io.appium.espressoserver.lib.model.Locator
+import io.appium.espressoserver.lib.model.toJsonMatcher
+import io.appium.espressoserver.lib.viewmatcher.withView
+import io.appium.espressoserver.lib.viewmatcher.withXPath
+import org.hamcrest.CoreMatchers.allOf
+import org.hamcrest.CoreMatchers.equalTo
+import org.hamcrest.CoreMatchers.`is`
+import org.hamcrest.Matchers.endsWith
+import org.hamcrest.Matchers.hasEntry
+import org.hamcrest.Matchers.instanceOf
+
+/**
+ * Helper methods to find elements based on locator strategies and selectors
+ */
+object ComposeNodeFinder {
+    fun findAllComposeNodeInteractions(
+        parentNode: SemanticsNodeInteraction?,
+        locator: Locator
+    ): SemanticsNodeInteractionCollection {
+        val matcher =
+            when (locator.using) {
+                Strategy.VIEW_TAG -> hasTestTag(locator.value!!)
+                Strategy.TEXT -> hasText(locator.value!!)
+                Strategy.LINK_TEXT -> hasText(locator.value!!)
+                Strategy.ACCESSIBILITY_ID -> hasContentDescription(locator.value!!)
+                else -> throw InvalidSelectorException(
+                    "Can't use non-Compose selectors. " +
+                            "only ${Strategy.VIEW_TAG}, ${Strategy.TEXT}, ${Strategy.LINK_TEXT} and " +
+                            "${Strategy.ACCESSIBILITY_ID} are supported"
+                )
+            }
+        return parentNode?.onChildren()?.filter(matcher)
```

Comment:
this looks a bit suspicious to me. We should not default to `onAllNodes` call if `onChildren()` or `filter` returns null

Reply:
I changed it to 
```
    return if (parentNode == null) {
        EspressoServerRunnerTest.composeTestRule.onAllNodes(matcher)
    } else {
        parentNode.onChildren().filter(matcher)
    }
```
But Android Studio still suggests changing it to elvis operator and auto-apply produces the code above. I left it as if-else anyway